### PR TITLE
Full Site Editing: Add a Site Logo block

### PR DESCRIPTION
**Pull request by scruffian**
Props @pablinos who wrote most of the code.

## Description
This adds a Site Logo block to the Full Site Editing experiment.

This uses some filters to non-destructively override the theme mod value when the experiment is enabled and the site logo setting has a value.

## How has this been tested?
- To test, enable the Full Site Editing experiment
- Add a template (e.g. https://github.com/WordPress/gutenberg/pull/18554)
- Add a Site Logo block to the template
- Add a Site Logo block to some content
- Try changing the alignment in the block toolbar
- Try modifying the width in the sidebar
- Try adding alt text in the media library
- Load the front end of the site and check that you can see the Site Logo

## Screenshots
Screenshot 2019-11-28 at 17 18 08
Screenshot 2019-11-28 at 17 17 29

## Known issues
If a non-admin triese to use this block they won't see anything.

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

**Comment by davemart-in** [comment]
Some quick feedback:

- Alignment worked as expected
- Loading the logo block after already uploading an image worked as expected (the logo was pre-loaded when I did it the second time)
- There's no logo icon next to "Site Logo" there in the middle (just under the logo in the screenshot above)
- Resizing the image didn't seem to work immediately after I uploaded the image. I had to delete the block then add it again to get the resizing to work. I believe this is due to the fact that it was in edit mode (See screenshot above). But I'm not sure this is easy to grok.
- The "Upload" option didn't work for me. I selected a file and nothing happened. I had to select "Media Library" and upload my logo there.
- The resizing option in the sidebar is still set to % instead of px (guessing this is still being worked on).

File: lib/template-loader.php
Comment (scruffian) [inline lib/template-loader.php]: I had to add this line to get templates working. Not sure what I'm doing wrong.

File: packages/block-library/src/site-logo/edit.js
Comment (scruffian) [inline packages/block-library/src/site-logo/edit.js]: Is there a way to hook into notices without using compose?

File: lib/template-loader.php
Comment (aduth) [reply lib/template-loader.php]: > I had to add this line to get templates working. Not sure what I'm doing wrong.

See also #18877. You may need to rebase.

**Comment by davemart-in** [comment]
Another round of testing. Here's what I found:

**Block logo**

- The solid icon style with the white star and the solid grey bg does not match up very well with the outline style of the other block icons (See: https://cl.ly/ef39e6725bc6). Can we swap this out for this SVG: https://d1czrtm2mp3lak.cloudfront.net/items/1R1D3s182R2i0v0S3I3e/logo.svg

**Uploading**
- I don't believe SVG files are allowed. When I try and upload one, nothing happens: https://cl.ly/38c8537baed6 Can we filter which file types are acceptable? The image block displays an inline error: https://cl.ly/e0cde373d136
- The image block shows a preview and a little loader icon immediately after you upload an image: https://cl.ly/242bfe94425e. The logo block just pauses for a long time and then appears once the file has been uploaded: https://cl.ly/fefa1ebbe185 Can we adopt the same loading state as the image block?

**Resizing:**
- Upon initial upload, the resize handles are not in the right place: https://cl.ly/1ba1478969da If you click then, they snap to the image. I noticed that if you upload the same image again, that this will not happen, so you'll need to use a new image. Here's the one I used: https://d1czrtm2mp3lak.cloudfront.net/items/3t1e2a1K392F3i1J2R1G/MyLogo.png
- Remove a logo block, then add it again and the logo dimensions are SUPER tiny: https://cl.ly/56df97dadc5f If you click the resize handle it magically resizes.
- When I edit a logo and select another image from the media library, the resize handles are all off. Clicking them does not help: https://cl.ly/a36aff132a54 My assumption is that they are stuck on the size of the previous logo.

**Alignment**
- Alignment within the image block seems to take place all within the bounds of the primary editor container: https://cl.ly/b181f5f6fd1a With the logo block the alignment options appear to take you outside of this container: https://cl.ly/1a5444bdaef8 It's not clear to me whether that was intentional.

**Duplicate**
- Creating a duplicate of the logo block creates a blank block container until you click outside, upon which the logo shows: https://cl.ly/78e6e5eb7927

**Edit as HTML**
- Edit as HTML produces no markup: https://cl.ly/63812a231d91 For comparison,. the image block produces the following markup: https://cl.ly/fb623fc4980f

**Reusable Block**
- Turning the logo block into a reusable block has some bugs: https://cl.ly/f855268e4513 At first it doesn't show up, until you click around. When you try and reuse it, the logo does not show up at all, despite clicking around. To contrast, here's what it looks like for the image block: https://cl.ly/e88bbbd27803

Hopefully that is helpful. Sorry for the big ole list! 😬

File: packages/block-library/src/site-logo/edit.js
Comment (scruffian) [inline packages/block-library/src/site-logo/edit.js]: This is just to facilitate testing, we'll remove it before merging.

**Comment by davemart-in** [comment]
Awesome work fixing everything up y'all. This is looking really great. One last super minor thing:

When you first add a logo block, the width dimension is not populated. You have to resize the logo for it to show. From then on, it stays visible:

hidden-dimensions

It's not a blocker, but it would be nice to show that dimension from the start.

And that's it. Again, nice work on this one!

**Comment by huguesvincent** [comment]
Congrats on this one. Will draft a call for testing today so that we can get it out when appropriate.

**Comment by mapk** [comment]
As indicated, this PR will also close #1851. What's the status here, @scruffian?

**Comment by scruffian** [comment]
Just waiting for a review

**Comment by scruffian** [comment]
@mtias @youknowriad what do you think about moving forward with this? I'll rebase it if you're on board...

File: packages/block-library/src/site-logo/edit.js
Comment (youknowriad) [inline packages/block-library/src/site-logo/edit.js:67]: @epiqueras Is this still the way to go for site entities?

File: packages/block-library/src/site-logo/edit.js
Comment (youknowriad) [inline packages/block-library/src/site-logo/edit.js:72]: Not sure why the saving is needed in the block itself, should we rely on the multi-entity saving flow instead (basically, just remove the saving code)

File: packages/block-library/src/site-logo/edit.js
Comment (youknowriad) [inline packages/block-library/src/site-logo/edit.js:88]: Any particular reason for using multiple useSelect calls instead of just one?

File: packages/block-library/src/site-logo/edit.js
Comment (youknowriad) [inline packages/block-library/src/site-logo/edit.js:283]: Feels like there's a few things in common between this block and the image  block. Not for this PR but it would be good to find a good way to share behavior.

File: packages/block-library/src/site-logo/icon.js
Comment (youknowriad) [inline packages/block-library/src/site-logo/icon.js:11]: Do you this icon can be added to @wordpress/icons or it's not generic enough?

File: packages/block-library/src/site-logo/index.js
Comment (youknowriad) [inline packages/block-library/src/site-logo/index.js:26]: Can we use the "align" hook to support alignments instead of having the logic in the block itself?

Comment (youknowriad) [reply packages/block-library/src/site-logo/index.js:26]: Maybe not that important for this block but can we try  using the lighterBlockDom flag too and the `Block.*` component?

File: packages/block-library/src/site-logo/index.php
Comment (youknowriad) [inline packages/block-library/src/site-logo/index.php:45]: I didn't check  but how close is the generated  markup here compared to the editor markup?

**Review by youknowriad — COMMENTED** [review]
Nice work here

File: packages/block-library/src/site-logo/edit.js
Comment (epiqueras) [reply packages/block-library/src/site-logo/edit.js:67]: No, see https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/post-title/edit.js.

It should be a small change.

**Comment by paaljoachim** [comment]
Hey Ben @scruffian 
It looks like this PR has almost reached the finish line. 
It will be great to have it merged. Thanks.

File: packages/block-library/src/site-logo/index.js
Comment (scruffian) [reply packages/block-library/src/site-logo/index.js:26]: What is the `lighterBlockDom` flag?

File: packages/block-library/src/site-logo/edit.js
Comment (scruffian) [reply packages/block-library/src/site-logo/edit.js:67]: `post-title/edit.js` doesn't seem to have an example for how to set a site entity, only how to fetch it.

Comment (epiqueras) [reply packages/block-library/src/site-logo/edit.js:67]: You can use `useDispatch( 'core' ).editEntityRecord`.

File: packages/block-library/src/site-logo/index.js
Comment (youknowriad) [reply packages/block-library/src/site-logo/index.js:26]: Here's an example of refactor to use that flag https://github.com/WordPress/gutenberg/pull/23062 
We still need to add some docs about it.

File: packages/block-library/src/site-logo/icon.js
Comment (scruffian) [reply packages/block-library/src/site-logo/icon.js:13]: Happy to add it if you think there's value in it. Probably best to do it in a followup PR.

File: packages/block-library/src/site-logo/edit.js
Comment (scruffian) [reply packages/block-library/src/site-logo/edit.js:227]: Removed

File: packages/block-library/src/site-logo/edit.js
Comment (scruffian) [reply packages/block-library/src/site-logo/edit.js:362]: Yes, I have extracted this to another component, but I think it would be good to share this code somehow...

File: packages/block-library/src/site-logo/edit.js
Comment (scruffian) [reply packages/block-library/src/site-logo/edit.js]: Done!

File: packages/block-library/src/site-logo/index.js
Comment (scruffian) [reply packages/block-library/src/site-logo/index.js:24]: Done

File: packages/block-library/src/site-logo/index.php
Comment (scruffian) [reply packages/block-library/src/site-logo/index.php:44]: I just made some changes to bring them into better alignment.

**Comment by scruffian** [comment]
@paaljoachim I think I've addressed all the concerns above, so this is ready for another review :)

File: packages/block-library/src/site-logo/edit.js
Comment (Copons) [inline packages/block-library/src/site-logo/edit.js:320]: I've tested on a site without logo, and when inserting the block, the placeholder shows up with an infinite spinner.

Screenshot 2020-06-23 at 18 40 02

If I add a logo and remove it, the spinner would (correctly) disappear.

File: packages/block-library/src/site-logo/edit.js
Comment (Copons) [inline packages/block-library/src/site-logo/edit.js:222]: I'm not much into `let`, and I think we could simplify some of them here with some early returns.
E.g.
```jsx
const img = <a href="#home" ... />;

if ( ! isResizable || ! imageWidthWithinContainer ) {
  return <div style={ { width, height } }>{ img }</div>;
}

return ( <>
  { <InspectorControls>{ ... }</InspectorControls> }
  { <ResizableBox>{ img }</ResizableBox> }
</> );
```

**Comment by pablohoneyhoney** [comment]
A more unique icon and more integrated with the latest library and styles might help with consistency and identification.
Perhaps a flag. 

block-logo

File: packages/block-library/src/site-logo/icon.js
Comment (scruffian) [inline packages/block-library/src/site-logo/icon.js:14]: We don't have `Line` as a primitive. Maybe we should add it?

**Comment by pablohoneyhoney** [comment]
An alt logo in case it holds the toolbar with a better hierarchy. 

block-logo2


The first one was

block-logo

File: packages/block-library/src/site-logo/edit.js
Comment (youknowriad) [inline packages/block-library/src/site-logo/edit.js:88]: Why `#home` as a href here? 

Minor: I find the separate "classes" variable to be just an unnecessary indirection. Inlining it might be better.

File: packages/block-library/src/site-logo/edit.js
Comment (youknowriad) [inline packages/block-library/src/site-logo/edit.js:89]: Why do we need this span?

File: packages/block-library/src/site-logo/edit.js
Comment (youknowriad) [inline packages/block-library/src/site-logo/edit.js:91]: Do we need a className here? is it for both frontend and editor or just for editor? In both cases, it doesn't follow the guidelines for any of these;

 - If it's frontend and editor it should be `wp-block-site-logo__something` 
 - if it's just for editor it should be `block-library-site-logo__something`

File: packages/block-library/src/site-logo/edit.js
Comment (youknowriad) [reply packages/block-library/src/site-logo/edit.js:243]: This question still stands.

File: packages/block-library/src/site-logo/editor.scss
Comment (youknowriad) [inline packages/block-library/src/site-logo/editor.scss:36]: I know it's not specific to this block, but altering styles of reusable components by targetting their classNames is not great. It's hard to find the right balance though

File: packages/block-library/src/site-logo/index.php
Comment (youknowriad) [inline packages/block-library/src/site-logo/index.php:40]: Should we remove this?

**Review by youknowriad — COMMENTED** [review]
Left some comments, this block is very interesting, it highlights the fact that we lack good abstractions to manipulate images without reusing the entire image block. Not something to solve in this PR but I would love some thoughts/follow-ups on this.

File: packages/block-library/src/site-logo/edit.js
Comment (scruffian) [reply packages/block-library/src/site-logo/edit.js:243]: They are separate because the call to get `mediaItemData` relies on `sitelogo` being loaded from the API, that's why `sitelogo` is passed as a second parameter to `useSelect`.  Is there another way of achieving that with only one `useSelect` call?

File: packages/block-library/src/site-logo/edit.js
Comment (scruffian) [reply packages/block-library/src/site-logo/edit.js:91]: The core function `get_custom_logo()` outputs this markup (both the span above and also this class name. I've replicated this in the editor so that the markup between the two is as close as possible.

Comment (youknowriad) [reply packages/block-library/src/site-logo/edit.js:91]: Oh, interesting! it makes sense to keep it as though.

File: packages/block-library/src/site-logo/edit.js
Comment (scruffian) [reply packages/block-library/src/site-logo/edit.js]: I updated the approach to output the actual site URL, and then added a click handler to stop it working. That way it will more closely match the output on the front end.

File: packages/block-library/src/site-logo/index.php
Comment (scruffian) [reply packages/block-library/src/site-logo/index.php:40]: ✅

File: packages/block-library/src/site-logo/editor.scss
Comment (scruffian) [reply packages/block-library/src/site-logo/editor.scss]: You're right, and I don't think we need this anymore anyways so I've taken it out.

File: packages/block-library/src/site-logo/edit.js
Comment (youknowriad) [reply packages/block-library/src/site-logo/edit.js]: I think there might be some a11y implications here though. We'd need some expert eyes.

File: packages/block-library/src/site-logo/edit.js
Comment (youknowriad) [reply packages/block-library/src/site-logo/edit.js:250]: I believe you can call the first selector, store the result on a local variable (inside mapSelect) and use it if it's available to perform the second call?

**Review by youknowriad — APPROVED** [review]
I didn't test a lot but it looks good to me.

File: packages/block-library/src/site-logo/edit.js
Comment (scruffian) [reply packages/block-library/src/site-logo/edit.js]: You're right, very nice! Done in 0d2585b.

File: packages/block-library/src/site-logo/edit.js
Comment (youknowriad) [inline packages/block-library/src/site-logo/edit.js:250]: Thanks for the update, I believe we should add a `[]` as a second argument here to ensure the callback is not executed on each render.

Comment (scruffian) [reply packages/block-library/src/site-logo/edit.js]: done in 2f08a3b530bfda2b52cf42a44be158a5652ac557

**Review by youknowriad — APPROVED** [review]

**Pull request merged**
